src/DataTypes/statepair_list.rs

```diff
@@ -0,0 +1,154 @@
+use std::collections::{HashMap, VecDeque};
+
+use crate::{
+    DBMLib::dbm::Federation, ModelObjects::statepair::StatePair,
+    TransitionSystems::transition_system::LocationID,
+};
+
+pub type PassedStateList = PassedStateListFed;
+type PassedStateListFed = HashMap<(LocationID, LocationID), Federation>;
+type PassedStateListVec = HashMap<(LocationID, LocationID), Vec<Federation>>;
+pub struct WaitingStateList {
+    queue: VecDeque<StatePair>,
+    map: HashMap<(LocationID, LocationID), VecDeque<Federation>>,
+}
+
+const BREADTH_FIRST: bool = false;
+
+pub trait PassedStateListExt {
+    fn put(&mut self, pair: StatePair);
+    fn has(&self, pair: &StatePair) -> bool;
+    fn zones(&self, key: &(LocationID, LocationID)) -> Vec<Federation>;
+}
+
+impl PassedStateListExt for PassedStateListVec {
+    fn put(&mut self, pair: StatePair) {
+        let (loc1, loc2, fed) = (pair.locations1.id, pair.locations2.id, pair.zone);
+        let key = (loc1, loc2);
+        if let Some(vec) = self.get_mut(&key) {
+            vec.push(fed);
+        } else {
+            self.insert(key, vec![fed]);
+        };
+    }
+
+    fn has(&self, pair: &StatePair) -> bool {
+        //let pair = pair.clone();
+        let (loc1, loc2, fed) = (
+            pair.locations1.id.clone(),
+            pair.locations2.id.clone(),
+            &pair.zone,
+        );
+        let key = (loc1, loc2);
+        match self.get(&key) {
+            Some(vec) => vec.iter().any(|f| fed.is_subset_eq(f)),
+            None => false,
+        }
```

Comment:
Unfortunately we cannot move `pair.locations1.id` because it is behind a shared reference. We could however implement copy for `LocationID` so we don't need to do the cloning everywhere, but i'm not sure what is most idiomatic

Reply:
Right, I had other errors hiding this one on my end. My bad. However, this reveals an issue with using tuples as keys on hashmap as these are clones are presumably very expensive.

I think Copy is only meant to be implemented on types where copying is trivial, so clone will do.